cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/DefaultSparkExecutionContext.scala

```diff
@@ -285,43 +282,47 @@ object DefaultSparkExecutionContext {
   private def createBatchWritableFunc[K, V]
       (datasetName: String,
        arguments: Map[String, String],
-       txServiceBaseURI: Broadcast[URI]) = (context: TaskContext, itor: Iterator[(K, V)]) => {
+       txServiceBaseURI: Broadcast[URI]) = (context: TaskContext, itor: Iterator
+    [(K, V)]) => {
 
-    val sparkTxClient = new SparkTransactionClient(txServiceBaseURI.value)
-    val datasetCache = SparkRuntimeContextProvider.get().getDatasetCache
-    val dataset: Dataset = datasetCache.getDataset(datasetName, arguments)
     val outputMetrics = new BatchWritableMetrics
-
     context.taskMetrics.outputMetrics = Option(outputMetrics)
 
-    // Creates an Option[TransactionAware] if the dataset is a TransactionAware
-    val txAware = dataset match {
-      case txAware: TransactionAware => Some(txAware)
-      case _ => None
-    }
+    val sparkTxClient = new SparkTransactionClient(txServiceBaseURI.value)
+    val datasetCache = SparkRuntimeContextProvider.get().getDatasetCache
+    val dataset: Dataset = datasetCache.getDataset(datasetName, arguments, true, AccessType.WRITE)
 
-    // Try to get the transaction for this stage. Hardcoded the timeout to 10 seconds for now
-    txAware.foreach(_.startTx(sparkTxClient.getTransaction(context.stageId(), 10, TimeUnit.SECONDS)));
-
-    // Write through BatchWritable.
-    val writable = dataset.asInstanceOf[BatchWritable[K, V]]
-    var records = 0;
-    while (itor.hasNext) {
-      val pair = itor.next()
-      writable.write(pair._1, pair._2)
-      outputMetrics.incrementRecordWrite(1)
-
-      // Periodically calling commitTx to flush changes. Hardcoded to 1000 records for now
-      if (records > 1000) {
-        txAware.foreach(_.commitTx())
-        records = 0
+    try {
+      // Creates an Option[TransactionAware] if the dataset is a TransactionAware
+      val txAware = dataset match {
+        case txAware: TransactionAware => Some(txAware)
+        case _ => None
       }
-      records += 1
-    }
 
-    // Flush all writes
-    txAware.foreach(_.commitTx())
-    datasetCache.releaseDataset(dataset)
+      // Try to get the transaction for this stage. Hardcoded the timeout to 10 seconds for now
+      txAware.foreach(_.startTx(sparkTxClient.getTransaction(context.stageId(), 10, TimeUnit.SECONDS)))
+
+      // Write through BatchWritable.
+      val writable = dataset.asInstanceOf[BatchWritable[K, V]]
+      var records = 0
+      while (itor.hasNext) {
+        val pair = itor.next()
+        writable.write(pair._1, pair._2)
+        outputMetrics.incrementRecordWrite(1)
+
+        // Periodically calling commitTx to flush changes. Hardcoded to 1000 records for now
```

Comment:
There is no JIRA covering this specific comment. However, the reason why we have to flush is because of the `BufferingTable` won't autoflush and the `BatchWritable` won't have an API to allow writing items in batch (and skip buffering). Fixing the underlying Dataset table hierarchy issue will have this code path changed anyway.